### PR TITLE
ace15: Add communication widget and used memory block reporting to PMC

### PR DESCRIPTION
**Pull request by softwarecki**
This PR will add Intel DSP Communication Widget driver for a device for generic sideband message transmit/receive between IPs in a SOC.
To MTL mm driver I add reporting of the number of HPSRAM memory blocks in use. This value is reported to the
PMC via the sideband channel supported by the communication widget.

### CODE REVIEW

File: drivers/misc/intel/comm_widget/intel_comm_widget.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		uint32_t iface = FIELD_PREP(CW_PMC_IPC_OP_CODE, CW_PMC_OPC_SRAM_CONFIG) |
+				 FIELD_PREP(CW_PMC_IPC_SRAM_USED_BANKS, self->memory_usage) |
+				 CW_PMC_IPC_BUSY;
+
+		self->sending = true;
+		cw_sb_write(CW_PMC_DESTID_VALUE, 0, CW_PMC_MAILBOX3_INTERFACE_ADDRESS, iface);
+
+		/* Wait for send interrupt */
+		k_sem_take(&self->tx_sem, K_FOREVER);
+		/* use cw_upstream_wait_for_sent() to wait in busy loop */
```
Comment (edersondisouza) [inline drivers/misc/intel/comm_widget/intel_comm_widget.c:96]: Is this comment some kind of TODO?

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 2]
```diff
@@ -64,15 +64,26 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
...
 
 #define L2_SRAM_PAGES_NUM			(L2_SRAM_SIZE / CONFIG_MM_DRV_PAGE_SIZE)
-#define MAX_EBB_BANKS_IN_SEGMENT	32
-#define SRAM_BANK_SIZE				(128 * 1024)
+#define MAX_EBB_BANKS_IN_SEGMENT		32
+#define SRAM_BANK_SIZE				KB(128)
 #define L2_SRAM_BANK_NUM			(L2_SRAM_SIZE / SRAM_BANK_SIZE)
-#define IS_BIT_SET(value, idx)		((value) & (1 << (idx)))
+#define IS_BIT_SET(value, idx)			((value) & (1 << (idx)))
+#define PMC_COMMINUCATION			defined(CONFIG_SOC_SERIES_INTEL_ACE1X) && \
```
Comment (edersondisouza) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:71]: Typo here? COMMUNICATION?

**Review by edersondisouza — CHANGES_REQUESTED** [review]

File: dts/bindings/sram/mmio-sram.yaml [hunk 3]
```diff
@@ -10,3 +10,7 @@ include: base.yaml
 properties:
     reg:
       required: true
+
+    pmc-communication:
```
Comment (galak) [inline dts/bindings/sram/mmio-sram.yaml:14]: What is this property and why are you adding it to the base `mmio-sram` binding?

**Review by galak — CHANGES_REQUESTED** [review]
`pmc-communication` is not a generic property that should be added to `mmio-sram`.  What is the purpose of this property?

**Review by galak — COMMENTED** [review]
Can we not use on of the existing IPC mechanism in Zephyr?  OpenAMP, mbox, ipc, etc?

File: drivers/misc/intel/comm_widget/intel_comm_widget.c (hunk 1 above)
Comment (softwarecki) [reply drivers/misc/intel/comm_widget/intel_comm_widget.c:96]: It is rather a note. Currently, the wait is based on the semaphore and the interrupt.

File: drivers/misc/intel/comm_widget/intel_comm_widget.c [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	events[0].state = K_POLL_STATE_NOT_READY;
+	while (1) {
+		k_poll(events, 1, K_FOREVER);
+		k_poll_signal_reset(&self->signal);
+		events[0].state = K_POLL_STATE_NOT_READY;
+
+		if (!cw_upstream_ready())
+			continue;
+
+		uint32_t iface = FIELD_PREP(CW_PMC_IPC_OP_CODE, CW_PMC_OPC_SRAM_CONFIG) |
```
[the remark this answers is not in the document]
Comment (softwarecki) [reply drivers/misc/intel/comm_widget/intel_comm_widget.c:87]: It is a missing third layer between tlb and communication widget. I will consider what to do with it. Maybe I will split it to a separate file.

**Comment by softwarecki** [comment]
> Can we not use on of the existing IPC mechanism in Zephyr? OpenAMP, mbox, ipc, etc?

It's a piece of SoC hardware. I suspect that it's based on https://patents.google.com/patent/US9213666B2/en .

File: drivers/misc/intel/comm_widget/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
+zephyr_library_sources(intel_comm_widget.c)
+zephyr_library_compile_options(-std=gnu99)
```
Comment (marc-hb) [inline drivers/misc/intel/comm_widget/CMakeLists.txt:6]: This does not "smell" good, why would such a very generic flag be buried deep down and apply to a single, brand new driver? What problem does this solve?

File: drivers/misc/intel/comm_widget/intel_comm_widget.c [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		__ASSERT(self->sending, "Spurious interrupt");
+
+		LOG_DBG("CW msgsent");
+		self->sending = false;
+		cw_upstream_clear_msgsent();
+		k_sem_give(&self->tx_sem);
+	}
+}
+
+static void comm_wdgt_thread(struct comm_wdgt_data *self)
```
[the remark this answers is not in the document]
Comment (andyross) [reply drivers/misc/intel/comm_widget/intel_comm_widget.c:69]: Actually, reading more carefully: it looks like you're not getting any async data from the PMC at all.  The only action is to wait for the other side to trigger an interrupt when the command was received, on which you want your calling code to block.  If that's the design, then I don't see why you need a thread or work item at all: just block the calling code (in an app thread somewhere) on a k_sem and wake it up in the ISR.  The threading architecture belongs in the app where possible, not the driver.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 7]
```diff
@@ -62,15 +62,26 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
...
 #define L2_SRAM_PAGES_NUM			(L2_SRAM_SIZE / CONFIG_MM_DRV_PAGE_SIZE)
-#define MAX_EBB_BANKS_IN_SEGMENT	32
-#define SRAM_BANK_SIZE				(128 * 1024)
+#define MAX_EBB_BANKS_IN_SEGMENT		32
+#define SRAM_BANK_SIZE				KB(128)
 #define L2_SRAM_BANK_NUM			(L2_SRAM_SIZE / SRAM_BANK_SIZE)
-#define IS_BIT_SET(value, idx)		((value) & (1 << (idx)))
+#define IS_BIT_SET(value, idx)			((value) & (1 << (idx)))
+#define PMC_COMMUNICATION			defined(CONFIG_SOC_SERIES_INTEL_ACE) && \
+						defined(CONFIG_MISC_INTEL_COMM_WIDGET)
```
Comment (andyross) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:70]: Put "depends on SOC_SERIES_INTEL_ACE" into the Kconfig definition for "MISC_INTEL_COMM_WIDGET" and you don't need this extra term in the test, and can just use the raw kconfig directly instead of inventing your own flag.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 8]
```diff
@@ -62,15 +62,26 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
...
 
 static struct k_spinlock tlb_lock;
 extern struct k_spinlock sys_mm_drv_common_lock;
 
 static int hpsram_ref[L2_SRAM_BANK_NUM];
+#if PMC_COMMUNICATION
+#include <zephyr/drivers/misc/intel/comm_widget.h>
+
+#define PMC_COMMUNICATION_DEV		DT_PHANDLE_BY_IDX(DT_NODELABEL(tlb), pmc_communication, 0)
```
Comment (andyross) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:79]: This seems backwards: you're assuming in C code, based on platform, that a DTS object must exist.  What you should be doing is checking to see if there is a DTS object (or better: a driver registered for it), whose configuration is set at the platform layer.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 9]
```diff
@@ -62,15 +62,26 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
...
+
+static uint32_t used_pages;
+static uint32_t used_pages_reported;
```
Comment (andyross) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:82]: Terminology nitpick: these are counts of 64k "banks", right?  Not "pages"?  Xtensa MMU support is going to show up eventually and it's going to be really confusing if we have two different kinds of pages.

**Review by andyross — COMMENTED** [review]
Some notes.  The big one is that I don't see that thread as having much value and think you should pick a leaner scheme.

**Comment by softwarecki** [comment]
I wonder if I should move this driver to soc/xtensa/intel_adsp/ace as it is platform specific. What do you think? @lgirdwood  @nashif @andyross @marc-hb @galak

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c (hunk 8 above)
Comment (softwarecki) [reply drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:79]: How I should check it?

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi [hunk 10]
```diff
@@ -430,5 +431,13 @@
 			lps = <&lps>;
 			bit-position = <13>;
 		};
+
+		ace_comm_widget: comm_widget@71C00 {
+			compatible = "intel,adsp-communication-widget";
+			reg = <0x00071C00 0x88>;
```
Comment (galak) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:437]: size should be 0x100

File: drivers/misc/intel/comm_widget/intel_comm_widget_messages.c [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "intel_comm_widget.h"
+#include "pmc_interface.h"
+
+/*
+ * Report number of used HP-SRAM memory banks to the PMC, unit is 32 KB.
```
Comment (galak) [inline drivers/misc/intel/comm_widget/intel_comm_widget_messages.c:15]: I'd replace 'Report' with 'Tell'.  Report made me think this is returning that value.

**Review by galak — CHANGES_REQUESTED** [review]
I don't think this needs to a "driver" at all.  This seems like it could just be some library code.

**Comment by galak** [comment]
> I wonder if I should move this driver to soc/xtensa/intel_adsp/ace as it is platform specific. What do you think? @lgirdwood @nashif @andyross @marc-hb @galak

I think its fine to move under soc/ and just make it some library code that utilizes information from devicetree.  don't see any need for this to be a full device/driver.  It will simplify `intel_comm_widget_pmc_send_ipc()` to not require a `struct device`.

**Comment by softwarecki** [comment]
This is a real device that can be used in many places for communication. At the moment it is only used in one place, so I gave up with handling interrupts, queueing and synchronization. We can do it in the future, when it becomes necessary to expand it, we will turn it into a real driver.

File: drivers/misc/intel/comm_widget/intel_comm_widget_messages.c [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "intel_comm_widget.h"
+#include "pmc_interface.h"
```
Comment (aborisovich) [inline drivers/misc/intel/comm_widget/intel_comm_widget_messages.c:12]: I think using relative paths is not a good thing.
Consider using \#include <zephyr/drivers/misc/intel_comm_widget/intel_comm_widget.h>
(just like other drivers do).
Or even better - move new headers to zephyr/include/zephyr/drivers/intel/comm_widget directory and include by
\#include <zephyr/drivers/intel/comm_widget/comm_widget.h>

File: drivers/misc/intel/comm_widget/intel_comm_widget.c [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+#include <stdint.h>
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+
+#define DT_DRV_COMPAT		intel_adsp_communication_widget
+
+#include "intel_comm_widget.h"
+
+#define SB_OPCODE_CRRD 0x06
+#define SB_OPCODE_CRWR 0x07
```
Comment (aborisovich) [inline drivers/misc/intel/comm_widget/intel_comm_widget.c:18]: What are those two?
Can you add some fancy /** doxygen explanation?

File: drivers/misc/intel/comm_widget/intel_comm_widget_messages.c [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+	if (!cw_upstream_ready())
+		return -EBUSY;
+
+	uint32_t iface = FIELD_PREP(CW_PMC_IPC_OP_CODE, CW_PMC_OPC_SRAM_CONFIG) |
+			 FIELD_PREP(CW_PMC_IPC_SRAM_USED_BANKS, banks) |
+			 CW_PMC_IPC_BUSY;
+
+	cw_sb_write(CW_PMC_DESTID_VALUE, 0, CW_PMC_MAILBOX3_INTERFACE_ADDRESS, iface);
+
+	cw_upstream_wait_for_sent();
```
Comment (aborisovich) [inline drivers/misc/intel/comm_widget/intel_comm_widget_messages.c:28]: Is this function defined anywhere? Can't find it in the code. Does this compile?
If it is not can you please add dummy functions that do nothing so it compiles?

File: drivers/misc/intel/comm_widget/intel_comm_widget.c [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+	cw_upstream_set_data(data);
+	cw_upstream_do_pw();
+}
+
+static int cw_init(const struct device *dev)
+{
+	return 0;
+}
+
+DEVICE_DT_DEFINE(CW_DT_NODE, cw_init, NULL, NULL, NULL, POST_KERNEL, 0, NULL);
```
Comment (aborisovich) [inline drivers/misc/intel/comm_widget/intel_comm_widget.c:33]: You use DT_DRV_COMPAT approach not NODELABEL.
DT_DRV_COMPAT expects you to define instance number (on the contrary to NODELABEL).
Please add required macros like in the guide https://docs.zephyrproject.org/latest/build/dts/howtos.html#option-1-create-devices-using-instance-numbers

Does cw_init function is mandatory? Can DEVICE_DT_DEFINE receive NULL there or will it fail to compile?

**Review by aborisovich — CHANGES_REQUESTED** [review]
Some refactoring is needed before this can be merged.
~I think we should move headers to zephyr/include directory where they belong~ (edit: those are implementation headers so this is ok), ~add macros required by DT_DRV_COMPAT (`#define CREATE_MY_DEVICE(inst)`)~ (I guess it is also not needed as your device does not use any structs for data and creation of many instances will just call empty init) ~add missing function definitions so the code compiles.~ (code compiles ok). Please address those remarks related device init function (it would be good to avoid adding empty function call just returning 0).

File: drivers/misc/intel/comm_widget/Kconfig [hunk 16]
```diff
@@ -0,0 +1,14 @@
+# Communication widget configuration options
+
+# Copyright (c) 2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_MISC_INTEL_COMM_WIDGET := intel,adsp-communication-widget
```
Comment (galak) [inline drivers/misc/intel/comm_widget/Kconfig:6]: remove `DT_COMPAT_MISC_INTEL_COMM_WIDGET` should get automatically created now.

File: drivers/misc/intel/comm_widget/Kconfig [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+
+config MISC_INTEL_COMM_WIDGET
+	bool "Intel Communication Widget driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_MISC_INTEL_COMM_WIDGET))
+	depends on SOC_SERIES_INTEL_ACE
```
Comment (galak) [inline drivers/misc/intel/comm_widget/Kconfig:11]: This should now be:

```
default y
depends on DT_HAS_INTEL_ADSP_COMMUNICATION_WIDGET_ENABLED
```

No need for `depends on SOC_SERIES_INTEL_ACE`.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 18]
```diff
@@ -46,13 +46,13 @@ DEVICE_MMIO_TOPLEVEL_STATIC(tlb_regs, DT_DRV_INST(0));
  * Number of significant bits in the page index (defines the size of
  * the table)
  */
-#define TLB_PADDR_SIZE DT_INST_PROP(0, paddr_size)
```
Comment (galak) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:49]: put reformatting, fixup in a separate commit and ideally a separate PR.

**Review by galak — CHANGES_REQUESTED** [review]
* Kconfig needs fixing up
* reformatting of mm_drv_intel_adsp_mtl_tlb.c should be in a separate commit/PR.

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi [hunk 19]
```diff
@@ -430,5 +431,13 @@
 			lps = <&lps>;
 			bit-position = <13>;
 		};
+
+		ace_comm_widget: comm_widget@71C00 {
```
Comment (galak) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:435]: Put this node in proper address sorted location.

File: drivers/misc/intel/comm_widget/intel_comm_widget_messages.c (hunk 14 above)
Comment (softwarecki) [reply drivers/misc/intel/comm_widget/intel_comm_widget_messages.c:28]: Yes, it is defined as static inline in drivers\misc\intel\comm_widget\intel_comm_widget.h. I always compile own code before pushing it as PR...

Comment (aborisovich) [reply drivers/misc/intel/comm_widget/intel_comm_widget_messages.c:28]: Good for you, just wanted to confirm :-)

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi [hunk 20]
```diff
@@ -313,6 +313,14 @@
 			};
 		};
 
+		ace_comm_widget: comm_widget@71C00 {
```
Comment (galak) [inline dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:316]: nit: `ace_comm_widget:` -> `comm_widget:`

File: dts/bindings/misc/intel,adsp-communication-widget.yaml [hunk 21]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2022 Intel Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+description: Intel, DSP Communication Widget
```
Comment (galak) [inline dts/bindings/misc/intel,adsp-communication-widget.yaml:4]: `description: Intel ADSP Communication Widget`

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi (hunk 20 above)
Comment (aborisovich) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi:316]: what does "nit" mean?
Please be more specific.

**Comment by aborisovich** [comment]
Some CI checks are still red.

**Review by aborisovich — COMMENTED** [review]
Looks good to me, but CI is still red - please fix.
And of course, compatible description should match its name like @galak already indicated.

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 22]
```diff
@@ -277,8 +297,12 @@ int sys_mm_drv_map_page(void *virt, uintptr_t phys, uint32_t flags)
 
 	entry_idx = get_tlb_entry_idx(va);
 
-	bank_idx = get_hpsram_bank_idx(pa);
+#ifdef CONFIG_SOC_INTEL_COMM_WIDGET
+	used_pages++;
```
Comment (abonislawski) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:301]: Why this is calculated?
PMC needs report about banks usage, in this code you already know when power on/off occurs, you can report to PMC in such cases

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c [hunk 23]
```diff
@@ -209,6 +215,20 @@ static int sys_mm_drv_hpsram_pwr(uint32_t bank_idx, bool enable, bool non_blocki
 	return 0;
 }
 
+#ifdef CONFIG_SOC_INTEL_COMM_WIDGET
+static void sys_mm_drv_report_page_usage(void)
+{
+	uint32_t banks = ceiling_fraction(used_pages, KB(32) / CONFIG_MM_DRV_PAGE_SIZE);
```
Comment (abonislawski) [inline drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:221]: Why not just use hpsram_ref with the real info about powered memory banks?

File: drivers/mm/mm_drv_intel_adsp_mtl_tlb.c (hunk 22 above)
Comment (softwarecki) [reply drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:301]: The power supply controls 128KB blocks, the use of 32kb blocks is reported to the PMC.

Comment (abonislawski) [reply drivers/mm/mm_drv_intel_adsp_mtl_tlb.c:301]: So this make sense now but it needs to be marked somewhere, in the same file we already have memory "banks" which means actually something else than newly added "banks".
At least please add comments but best to use different names everywhere where this is different bank, maybe pmc_bank? and comment to used_banks_reported what this means

File: dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi (hunk 20 above)
Comment (galak) [reply dts/xtensa/intel/intel_adsp_ace15_mtpm.dtsi]: Sorry, nit is short for nitpick.  (nitpick definition is: engage in fussy or pedantic fault-finding).  So using it to say that the comment is a pretty minor detail.

File: soc/xtensa/intel_adsp/ace/Kconfig.soc [hunk 24]
```diff
@@ -9,3 +9,10 @@ choice
 		depends on SOC_SERIES_INTEL_ACE
 
 endchoice
+
+config SOC_INTEL_COMM_WIDGET
+	bool "Intel Communication Widget driver"
+	default DT_HAS_INTEL_ADSP_COMMUNICATION_WIDGET_ENABLED
```
Comment (galak) [inline soc/xtensa/intel_adsp/ace/Kconfig.soc:15]: This should be:
```
default y
depends on DT_HAS_INTEL_ADSP_COMMUNICATION_WIDGET_ENABLED
```

File: soc/xtensa/intel_adsp/ace/comm_widget.c [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "comm_widget.h"
+
+/*
+ * Operation codes
+ */
```
Comment (galak) [inline soc/xtensa/intel_adsp/ace/comm_widget.c:11]: what does SB stand for here?

**Review by galak — CHANGES_REQUESTED** [review]
A few comments, also can we be consistent with naming being it 'cw' or 'comm_widget' but not both.  I'd say 'cw'.

File: soc/xtensa/intel_adsp/ace/comm_widget.c (hunk 25 above)
Comment (softwarecki) [reply soc/xtensa/intel_adsp/ace/comm_widget.c:11]: SideBand

File: soc/xtensa/intel_adsp/ace/comm_widget.h [hunk 26]
```diff
@@ -0,0 +1,863 @@
...
+	sts |= USSTS_MSGSENT;
+	sys_write32(sts, CW_BASE + USSTS);
+}
+
+/*
+ * @brief Wait for message to be send.
+ */
+static inline void cw_upstream_wait_for_sent(void)
+{
+	while ((sys_read32(CW_BASE + USSTS) & USSTS_MSGSENT) == 0) {
```
Comment (abonislawski) [inline soc/xtensa/intel_adsp/ace/comm_widget.h:851]: WAIT_FOR

**Comment by softwarecki** [comment]
@galak 
All internal functions used by driver was named cw_*. Only the one function (adsp_comm_widget_pmc_send_ipc) was named differently. It is exported outside the driver and used by another driver. I prefer keep this name, because cw is meaningless in tlb driver.

**Comment by softwarecki** [comment]
@galak @aborisovich @edersondisouza @abonislawski Please review

**Review by abonislawski — DISMISSED** [review]

**Review by aborisovich — DISMISSED** [review]

**Review by aborisovich — DISMISSED** [review]

**Review by abonislawski — DISMISSED** [review]

**Review by tmleman — DISMISSED** [review]

**Review by jxstelter — DISMISSED** [review]

**Review by edersondisouza — DISMISSED** [review]

**Comment by softwarecki** [comment]
@galak Please review

**Review by galak — DISMISSED** [review]

**Comment by nashif** [comment]
@softwarecki please rebase, there is a conflict.

**Review by abonislawski — DISMISSED** [review]

**Review by pjdobrowolski — DISMISSED** [review]

**Review by abonislawski — DISMISSED** [review]

**Review by marcinszkudlinski — DISMISSED** [review]

**Review by aborisovich — DISMISSED** [review]

**Review by nashif — APPROVED** [review]

**Review by aborisovich — APPROVED** [review]

**Review by abonislawski — APPROVED** [review]
